src/reflect/scala/reflect/internal/Symbols.scala

```diff
@@ -2371,23 +2371,35 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
       else this :: Nil
     )
 
+    private[this] var overridingCache: List[Symbol] = Nil
+    private[this] var overridingPeriod: Period = NoPeriod
+
     /** Returns all symbols overridden by this symbol. */
     final def allOverriddenSymbols: List[Symbol] = {
-      if (isOverridingSymbol) {
-        // performance sensitive
-        val builder = List.newBuilder[Symbol]
-        for (o <- owner.ancestors) {
-          overriddenSymbol(o).andAlso(builder += _)
-        }
-        builder.result()
-      } else Nil
+      if (overridingPeriod == currentPeriod) overridingCache
+      else {
+        val result = if (canMatchInheritedSymbols) {
+          // performance sensitive
+          val builder = List.newBuilder[Symbol]
+          for (o <- owner.ancestors) {
+            overriddenSymbol(o).andAlso(builder += _)
+          }
+          builder.result()
+        } else Nil
+        overridingPeriod = currentPeriod
+        overridingCache = result
+        result
+      }
```

Comment:
I'm reluctant to introduce this cache. This method is called on each decl in each class symbol (source or classpath based) during specialization, and these caches will hang around until the end of the compiler. It appears as though the caches would also consume O(N^2) space wrt the length of the override chain (the N-th deepest symbol would have a cache of N-1 elements).

I forget the exact reason that the cache in `isOverridingSymbol` got in the way of this bugfix, could you refresh my memory? My changes to `isOverridingSymbol` in #6197 will conflict here.